src/common/webrequest_curl.cpp

```diff
@@ -0,0 +1,559 @@
+{
+    DestroyHeaderList();
+
+    curl_easy_cleanup(m_handle);
+}
+
+void wxWebRequestCURL::Start()
+{
+    if ( GetState() != State_Idle )
+        return;
+
+    m_response.reset(new wxWebResponseCURL(*this));
+    m_response->Init();
+
+    if ( m_dataSize )
+    {
+        if ( m_method.empty() || m_method.IsSameAs("POST", false) )
+        {
+            curl_easy_setopt(m_handle, CURLOPT_POSTFIELDSIZE_LARGE,
+                static_cast<curl_off_t>(m_dataSize));
+            curl_easy_setopt(m_handle, CURLOPT_POST, 1L);
+        }
+        else if ( m_method.IsSameAs("PUT", false) )
+        {
+            curl_easy_setopt(m_handle, CURLOPT_UPLOAD, 1L);
+            curl_easy_setopt(m_handle, CURLOPT_INFILESIZE_LARGE,
+                static_cast<curl_off_t>(m_dataSize));
+        }
+    }
+
+    if ( m_method.IsSameAs("HEAD", false) )
+    {
+        curl_easy_setopt(m_handle, CURLOPT_NOBODY, 1L);
+    }
+    else if ( !m_method.empty() )
+    {
+        curl_easy_setopt(m_handle, CURLOPT_CUSTOMREQUEST,
+            static_cast<const char*>(m_method.mb_str()));
+    }
+
+    for ( wxWebRequestHeaderMap::const_iterator it = m_headers.begin();
+        it != m_headers.end(); ++it )
+    {
+        wxString hdrStr = wxString::Format("%s: %s", it->first, it->second);
+        m_headerList = curl_slist_append(m_headerList, hdrStr.mb_str());
+    }
+    curl_easy_setopt(m_handle, CURLOPT_HTTPHEADER, m_headerList);
+
+    StartRequest();
+}
+
+bool wxWebRequestCURL::StartRequest()
+{
+    m_bytesSent = 0;
+
+    if ( static_cast<wxWebSessionCURL&>(GetSession()).StartRequest(*this) )
+    {
+        SetState(State_Active);
+        return true;
+    }
+    else
+    {
+        SetState(State_Failed);
+        return false;
+    }
+}
+
+void wxWebRequestCURL::Cancel()
+{
+    static_cast<wxWebSessionCURL&>(GetSession()).CancelRequest(this);
+}
+
+void wxWebRequestCURL::HandleCompletion()
+{
+    int status = (m_response) ? m_response->GetStatus() : 0;
+
+    if ( status == 0)
+        SetState(State_Failed, GetError());
+    else if ( status == 401 || status == 407 )
+    {
+        m_authChallenge.reset(new wxWebAuthChallengeCURL(
+            (status == 407) ? wxWebAuthChallenge::Source_Proxy : wxWebAuthChallenge::Source_Server, *this));
+        if ( m_authChallenge->Init() )
+            SetState(State_Unauthorized, m_response->GetStatusText());
+        else
+            SetState(State_Failed);
+    }
+    else if ( CheckServerStatus() )
+        SetState(wxWebRequest::State_Completed);
+}
+
+wxString wxWebRequestCURL::GetError() const
+{
+    return wxString(m_errorBuffer);
+}
+
+size_t wxWebRequestCURL::ReadData(char* buffer, size_t size)
+{
+    if ( m_dataStream )
+    {
+        m_dataStream->Read((void*)buffer, size);
+        size_t readSize = m_dataStream->LastRead();
+        m_bytesSent += readSize;
+        return readSize;
+    }
+    else
+        return 0;
+}
+
+void wxWebRequestCURL::DestroyHeaderList()
```

Comment:
I wonder why extract this into a separate function when it's called only from dtor?